x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/event_log/get_execution_event_aggregation/index.ts

```diff
@@ -0,0 +1,351 @@
+import { BadRequestError } from '@kbn/securitysolution-es-utils';
+import { flatMap, get } from 'lodash';
+import { MAX_EXECUTION_EVENTS_DISPLAYED } from '@kbn/securitysolution-rules';
+import { AggregateEventsBySavedObjectResult } from '../../../../../../../event_log/server';
+import { AggregateRuleExecutionEvent } from '../../../../../../common/detection_engine/schemas/common';
+import { GetAggregateRuleExecutionEventsResponse } from '../../../../../../common/detection_engine/schemas/response';
+import {
+  ExecutionEventAggregationOptions,
+  ExecutionUuidAggResult,
+  ExecutionUuidAggBucket,
+} from './types';
+
+// Base ECS fields
+const ACTION_FIELD = 'event.action';
+const DURATION_FIELD = 'event.duration';
+const MESSAGE_FIELD = 'message';
+const PROVIDER_FIELD = 'event.provider';
+const OUTCOME_FIELD = 'event.outcome';
+const START_FIELD = 'event.start';
+const TIMESTAMP_FIELD = '@timestamp';
+// Platform fields
+const SCHEDULE_DELAY_FIELD = 'kibana.task.schedule_delay';
+const ES_SEARCH_DURATION_FIELD = 'kibana.alert.rule.execution.metrics.es_search_duration_ms';
+const TOTAL_ACTIONS_TRIGGERED_FIELD =
+  'kibana.alert.rule.execution.metrics.number_of_triggered_actions';
+const EXECUTION_UUID_FIELD = 'kibana.alert.rule.execution.uuid';
+// TODO: To be added in https://github.com/elastic/kibana/pull/126210
+// const TOTAL_ALERTS_CREATED: 'kibana.alert.rule.execution.metrics.total_alerts_created',
+// const TOTAL_ALERTS_DETECTED: 'kibana.alert.rule.execution.metrics.total_alerts_detected',
+// Security fields
+const GAP_DURATION_FIELD = 'kibana.alert.rule.execution.metrics.execution_gap_duration_s';
+const INDEXING_DURATION_FIELD = 'kibana.alert.rule.execution.metrics.total_indexing_duration_ms';
+const SEARCH_DURATION_FIELD = 'kibana.alert.rule.execution.metrics.total_search_duration_ms';
+const STATUS_FIELD = 'kibana.alert.rule.execution.status';
+
+const ONE_MILLISECOND_AS_NANOSECONDS = 1_000_000;
+
+const SORT_FIELD_TO_AGG_MAPPING: Record<string, string> = {
+  timestamp: 'ruleExecution>executeStartTime',
+  duration_ms: 'ruleExecution>executionDuration',
+  indexing_duration_ms: 'securityMetrics>indexDuration',
+  search_duration_ms: 'securityMetrics>searchDuration',
+  gap_duration_ms: 'securityMetrics>gapDuration',
+  schedule_delay_ms: 'ruleExecution>scheduleDelay',
+  num_triggered_actions: 'ruleExecution>numTriggeredActions',
+  // TODO: To be added in https://github.com/elastic/kibana/pull/126210
+  // total_alerts_created: 'securityMetrics>totalAlertsDetected',
+  // total_alerts_detected: 'securityMetrics>totalAlertsCreated',
+};
+
+/**
+ * Returns `aggs` to be supplied to aggregateEventsBySavedObjectIds
+ * @param maxExecutions upper bounds of execution events to return (to narrow below max terms agg limit)
+ * @param page current page to retrieve, starting at 0
+ * @param perPage number of execution events to display per page
+ * @param sort field to sort on
+ */
+export const getExecutionEventAggregation = ({
```

Comment:
Should we also pull out errors from the Alerting framework with this aggregation?

I have a rule that failed. The latest message in the event log is:

```json
{
  "@timestamp": "2022-03-24T14:00:38.873Z",
  "event": {
    "provider": "alerting",
    "action": "execute",
    "kind": "alert",
    "category": ["siem"],
    "start": "2022-03-24T14:00:38.868Z",
    "end": "2022-03-24T14:00:38.873Z",
    "duration": 5000000,
    "reason": "disabled",
    "outcome": "failure"
  },
  "kibana": { ... },
  "rule": { ... },
  "error": {
    "message": "Rule failed to execute because rule ran after it was disabled."
  },
  "message": "siem.queryRule:054e9470-a52d-11ec-9c80-0db5a784ac2c: execution failed"
}
```

However, a row with only start time is displayed on the execution log table:

<img width="2251" alt="Screenshot 2022-03-24 at 15 01 20" src="https://user-images.githubusercontent.com/1938181/159936086-544b8a9b-c52f-4041-aa27-cc7771445ce7.png">

Would it be possible to output the rule status and error message in the table? Similar to how we do that in the rule header:

<img width="647" alt="Screenshot 2022-03-24 at 15 20 01" src="https://user-images.githubusercontent.com/1938181/159937298-62df5554-0c9e-4e3b-abb3-de4eb657468c.png">

Reply:
Oooh, this is interesting! How can I reproduce this error? Just manually disable a rule right before it's scheduled to run? If it's not easy to repro, can you please share the JSON docs for that specific execution so I can load and test locally?

And I had thought about just concatenating all stack errors from the `alerting-execute` document as well, but it proved to be more noise and often not as useful as the `message` we're writing on `status-change` docs. That said, let me see what I can do for this error type so we can at least show `status` and `message` like we do in the header.

Reply:
I had an active rule that was scheduled to run every second, deactivated it, and received that error immediately. Probably we could see something similar if a rule gets canceled or in case of any other framework-level error 🤔

Reply:
Here are the last five events for that rule, just in case:

```json
{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 480,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [
      {
        "_index" : ".kibana-event-log-8.2.0-000001",
        "_id" : "70s6vH8BXr5aLARdtrrg",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2022-03-24T14:00:38.873Z",
          "event" : {
            "provider" : "alerting",
            "action" : "execute",
            "kind" : "alert",
            "category" : [
              "siem"
            ],
            "start" : "2022-03-24T14:00:38.868Z",
            "end" : "2022-03-24T14:00:38.873Z",
            "duration" : 5000000,
            "reason" : "disabled",
            "outcome" : "failure"
          },
          "kibana" : {
            "alert" : {
              "rule" : {
                "execution" : {
                  "uuid" : "3e177cd7-2225-430d-b33a-8b30c36d9ee8"
                }
              }
            },
            "saved_objects" : [
              {
                "rel" : "primary",
                "type" : "alert",
                "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
                "type_id" : "siem.queryRule"
              }
            ],
            "task" : {
              "scheduled" : "2022-03-24T14:00:36.796Z",
              "schedule_delay" : 2072000000
            },
            "alerting" : {
              "status" : "error"
            },
            "server_uuid" : "5b2de169-2785-441b-ae8c-186a1936b17d",
            "version" : "8.2.0"
          },
          "rule" : {
            "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
            "license" : "basic",
            "category" : "siem.queryRule",
            "ruleset" : "siem"
          },
          "error" : {
            "message" : "Rule failed to execute because rule ran after it was disabled."
          },
          "message" : "siem.queryRule:054e9470-a52d-11ec-9c80-0db5a784ac2c: execution failed",
          "ecs" : {
            "version" : "1.8.0"
          }
        },
        "sort" : [
          1648130438873
        ]
      },
      {
        "_index" : ".kibana-event-log-8.2.0-000001",
        "_id" : "7ks6vH8BXr5aLARdtrrg",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2022-03-24T14:00:38.868Z",
          "event" : {
            "provider" : "alerting",
            "action" : "execute-start",
            "kind" : "alert",
            "category" : [
              "siem"
            ],
            "start" : "2022-03-24T14:00:38.868Z"
          },
          "kibana" : {
            "alert" : {
              "rule" : {
                "execution" : {
                  "uuid" : "3e177cd7-2225-430d-b33a-8b30c36d9ee8"
                }
              }
            },
            "saved_objects" : [
              {
                "rel" : "primary",
                "type" : "alert",
                "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
                "type_id" : "siem.queryRule"
              }
            ],
            "task" : {
              "scheduled" : "2022-03-24T14:00:36.796Z",
              "schedule_delay" : 2072000000
            },
            "server_uuid" : "5b2de169-2785-441b-ae8c-186a1936b17d",
            "version" : "8.2.0"
          },
          "rule" : {
            "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
            "license" : "basic",
            "category" : "siem.queryRule",
            "ruleset" : "siem"
          },
          "message" : "rule execution start: \"054e9470-a52d-11ec-9c80-0db5a784ac2c\"",
          "ecs" : {
            "version" : "1.8.0"
          }
        },
        "sort" : [
          1648130438868
        ]
      },
      {
        "_index" : ".kibana-event-log-8.2.0-000001",
        "_id" : "7Us6vH8BXr5aLARdsrr4",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2022-03-24T14:00:37.411Z",
          "event" : {
            "provider" : "alerting",
            "action" : "execute",
            "kind" : "alert",
            "category" : [
              "siem"
            ],
            "start" : "2022-03-24T14:00:35.891Z",
            "outcome" : "success",
            "end" : "2022-03-24T14:00:37.411Z",
            "duration" : 1520000000
          },
          "kibana" : {
            "alert" : {
              "rule" : {
                "execution" : {
                  "uuid" : "e5b59b4a-94b7-4aa3-8e33-72c8e1a11585",
                  "metrics" : {
                    "number_of_triggered_actions" : 0,
                    "number_of_searches" : 3,
                    "es_search_duration_ms" : 0,
                    "total_search_duration_ms" : 3
                  }
                }
              }
            },
            "saved_objects" : [
              {
                "rel" : "primary",
                "type" : "alert",
                "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
                "type_id" : "siem.queryRule"
              }
            ],
            "task" : {
              "scheduled" : "2022-03-24T14:00:33.765Z",
              "schedule_delay" : 2126000000
            },
            "alerting" : {
              "status" : "ok"
            },
            "server_uuid" : "5b2de169-2785-441b-ae8c-186a1936b17d",
            "version" : "8.2.0"
          },
          "rule" : {
            "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
            "license" : "basic",
            "category" : "siem.queryRule",
            "ruleset" : "siem",
            "name" : "Test"
          },
          "message" : "rule executed: siem.queryRule:054e9470-a52d-11ec-9c80-0db5a784ac2c: 'Test'",
          "ecs" : {
            "version" : "1.8.0"
          }
        },
        "sort" : [
          1648130437411
        ]
      },
      {
        "_index" : ".kibana-event-log-8.2.0-000001",
        "_id" : "7Es6vH8BXr5aLARdr7oQ",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2022-03-24T14:00:36.394Z",
          "event" : {
            "provider" : "securitySolution.ruleExecution",
            "kind" : "event",
            "action" : "status-change",
            "sequence" : 1
          },
          "message" : """This rule is attempting to query data from Elasticsearch indices listed in the "Index pattern" section of the rule definition, however no index matching: ["apm-*-transaction*","traces-apm*","auditbeat-*","endgame-*","filebeat-*","logs-*","packetbeat-*","winlogbeat-*"] was found. This warning will continue to appear until a matching index is created or this rule is disabled.""",
          "rule" : {
            "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
            "name" : "Test",
            "category" : "siem.queryRule"
          },
          "kibana" : {
            "alert" : {
              "rule" : {
                "execution" : {
                  "status" : "partial failure",
                  "status_order" : 20,
                  "uuid" : "e5b59b4a-94b7-4aa3-8e33-72c8e1a11585"
                }
              }
            },
            "space_ids" : [
              "default"
            ],
            "saved_objects" : [
              {
                "rel" : "primary",
                "type" : "alert",
                "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c"
              }
            ],
            "server_uuid" : "5b2de169-2785-441b-ae8c-186a1936b17d",
            "version" : "8.2.0"
          },
          "ecs" : {
            "version" : "1.8.0"
          }
        },
        "sort" : [
          1648130436394
        ]
      },
      {
        "_index" : ".kibana-event-log-8.2.0-000001",
        "_id" : "60s6vH8BXr5aLARdr7oQ",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2022-03-24T14:00:35.910Z",
          "event" : {
            "provider" : "securitySolution.ruleExecution",
            "kind" : "event",
            "action" : "status-change",
            "sequence" : 0
          },
          "message" : "",
          "rule" : {
            "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c",
            "name" : "Test",
            "category" : "siem.queryRule"
          },
          "kibana" : {
            "alert" : {
              "rule" : {
                "execution" : {
                  "status" : "running",
                  "status_order" : 15,
                  "uuid" : "e5b59b4a-94b7-4aa3-8e33-72c8e1a11585"
                }
              }
            },
            "space_ids" : [
              "default"
            ],
            "saved_objects" : [
              {
                "rel" : "primary",
                "type" : "alert",
                "id" : "054e9470-a52d-11ec-9c80-0db5a784ac2c"
              }
            ],
            "server_uuid" : "5b2de169-2785-441b-ae8c-186a1936b17d",
            "version" : "8.2.0"
          },
          "ecs" : {
            "version" : "1.8.0"
          }
        },
        "sort" : [
          1648130435910
        ]
      }
    ]
  }
}
```

Reply:
> I had an active rule that was scheduled to run every second, deactivated it, and received that error immediately. Probably we could see something similar if a rule gets canceled due to timeout or in case of any other framework-level error 🤔

Ahhh, nice! I thought they introduced a 1min minimum execution schedule limit recently, but I was able to update and repro! Thanks @xcrzx -- will see what I can do here to surface this error better! 👍 🙂 

<img width="1587" alt="image" src="https://user-images.githubusercontent.com/2946766/159975180-2ccd0361-33ed-4941-8e21-becebfbf473a.png">

Reply:
> Note: we're also capturing rule timeouts in the aggs query, but just not surfacing it in the UI as it was getting crowded. We can probably expose this along with alerts detected/created when I add support for enabling individual columns at a time? Would be nice to have it represented here though....could always write status as `timeout` for that event maybe?

Hmm, I would expect to see timeout as an error with the rule status to be `Failed` because the rule is likely doesn't write alerts even if it detects something. So from a user's perspective, it is not different from any other rule error. Is it something that we can support easily?

Reply:
Yeah, I'll probably do something similar as with the previous error and just override the status/message in the event renderer -- that should be fine since those columns aren't sortable.

Reply:
This is resolved in [3623419](https://github.com/elastic/kibana/pull/130072/commits/36234191468045dfeb5b9049a6996467538b3f91). Ended up modifying the aggs query to be more robust in returning execution events when there are only platform documents (no security docs). Updated `getAggregateExecutionEvents` to fallback to platform status from `event.outcome` if `security_status` is empty, and also falls back to `error.message` is `security_message` is empty. This also now queries for corresponding `event.outcome` if a filter is provided so that platform-only events can still be displayed when filtering. 

That means this `rule disabled while running` error will now show both relevant status & message fields, and also be shown when global filters are applied.



![image](https://user-images.githubusercontent.com/2946766/164513694-b77af6d6-5cbd-4824-9155-79074e882a53.png)